internal/analyzer/analyzer.go

```diff
@@ -2,45 +2,95 @@ package analyzer
 
 import (
 	"context"
+	"fmt"
+	"hash/fnv"
+	"os"
+	"path/filepath"
 
+	"google.golang.org/protobuf/proto"
+
+	"github.com/sqlc-dev/sqlc/internal/analyzer/pb"
+	"github.com/sqlc-dev/sqlc/internal/cache"
+	"github.com/sqlc-dev/sqlc/internal/config"
+	"github.com/sqlc-dev/sqlc/internal/info"
 	"github.com/sqlc-dev/sqlc/internal/sql/ast"
 	"github.com/sqlc-dev/sqlc/internal/sql/named"
 )
 
-type Column struct {
-	Name         string
-	OriginalName string
-	DataType     string
-	NotNull      bool
-	Unsigned     bool
-	IsArray      bool
-	ArrayDims    int
-	Comment      string
-	Length       *int
-	IsNamedParam bool
-	IsFuncCall   bool
-
-	// XXX: Figure out what PostgreSQL calls `foo.id`
-	Scope      string
-	Table      *ast.TableName
-	TableAlias string
-	Type       *ast.TypeName
-	EmbedTable *ast.TableName
-
-	IsSqlcSlice bool // is this sqlc.slice()
+type CachedAnalyzer struct {
+	a          Analyzer
+	configPath string
+	config     []byte
+	db         config.Database
+}
+
+func Cached(a Analyzer, cp string, db config.Database) *CachedAnalyzer {
+	return &CachedAnalyzer{
+		a:          a,
+		configPath: cp,
+		db:         db,
+	}
 }
 
-type Parameter struct {
-	Number int
-	Column *Column
+func (c *CachedAnalyzer) Analyze(ctx context.Context, n ast.Node, q string, schema []string, np *named.ParamSet) (*pb.Analysis, error) {
+	// Only cache queries for managed databases. We can't be certain the the
+	// database is in an unchanged state otherwise
+	if !c.db.Managed {
+		return c.a.Analyze(ctx, n, q, schema, np)
+	}
+
+	dir, err := cache.AnalysisDir()
+	if err != nil {
+		return nil, err
+	}
+
+	if c.config == nil {
+		c.config, err = os.ReadFile(c.configPath)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
What about just serializing the `config.Config` that we've already constructed from the parsed config file bytes? You could store that on the struct instead of just `config.Database`.

That would have the advantage of creating consistent keys based on our internal config representation rather than the input. But maybe that's a disadvantage? I guess I'm thinking that if someone just adds a newline to their config that shouldn't really break the cached analysis.

Reply:
Yeah, that's a good idea. I wasn't trying to figure out how to serialize that struct and forgot that we serialize it to JSON / YAML already.